src/eventmaster/api_handlers.go

```diff
@@ -414,37 +414,43 @@ func (h *httpHandler) handleGetDc(w http.ResponseWriter, r *http.Request, _ http
 		h.sendError(w, http.StatusInternalServerError, err, "Error marshalling response to JSON", r.URL.Path)
 		return
 	}
-	h.sendResp(w, "", string(str), r.URL.Path)
+	h.sendResp(w, str, r.URL.Path)
 }
 
-func (h *httpHandler) handleGitHubEvent(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
-	var info map[string]interface{}
-
-	defer r.Body.Close()
-	reqBody, err := ioutil.ReadAll(r.Body)
-	if err != nil {
-		h.sendError(w, http.StatusBadRequest, err, "Error reading request body", r.URL.Path)
-		return
-	}
-	if err := json.Unmarshal(reqBody, &info); err != nil {
-		h.sendError(w, http.StatusBadRequest, err, "Error JSON decoding body of request", r.URL.Path)
-		return
-	}
-
-	id, err := h.store.AddEvent(&UnaddedEvent{
-		Dc:        "github",
-		Host:      "github",
-		TopicName: "github",
-		Data:      info,
-	})
-	if err != nil {
-		h.sendError(w, http.StatusInternalServerError, err, "Error adding event to store", r.URL.Path)
-		return
+func (h *httpHandler) handlePluginEvent(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	pluginName := ps.ByName("name")
+	if plugin, ok := Plugins[pluginName]; ok {
+		defer r.Body.Close()
+		rawData, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			h.sendError(w, http.StatusInternalServerError, err, "Error reading request body", r.URL.Path)
+			return
+		}
+		valid := plugin.ValidateAuth(r, rawData)
+		if !valid {
+			h.sendError(w, http.StatusUnauthorized, errors.New("Invalid signature"), "Error validating request", r.URL.Path)
+			return
+		}
+		events, err := plugin.ParseRequest(r, rawData)
+		if err != nil {
+			h.sendError(w, http.StatusBadRequest, err, "Error parsing event", r.URL.Path)
+			return
+		}
+		var ids []string
+		for _, event := range events {
+			id, err := h.store.AddEvent(event)
+			if err != nil {
+				h.sendError(w, http.StatusInternalServerError, err, "Error adding event to store", r.URL.Path)
```

Comment:
This will cause the first event send to fail the entire request. This is a problem because some events may have fired, but the caller doesn't know which ones. This should either be transactional (which is probably more work than its worth) or it should return a list of responses (which events it sent etc.)